Bz1959336

lnykryn opened the pull request.

pmgdeb committed:
Fixed some SUSE specific typos and outputs

(cherry picked from commit f18d069d4664c148df8563139b358806ecfe6799)

Resolves: #1959336

ryncsn committed:
squash: unsquash the root image instead of mounting it on shutdown

When building squash image, squash module forgot to install the new
shutdown.sh, and the shutdown hooks are always skipped on ordinary
shutdown if squash module is enabled.

The new shutdown.sh will remount the squash image and then everything
will just work, but currently re-mounting the squash image on shutdown
may have selinux problem and make the system hang, and there is no
easy way to fix it.

So skip fixing the shutdown.sh not being install problem, instead
just drop the new shutdown.sh, and unsquash the image on ordinary
shutdown, which is safer and should always work.

Signed-off-by: Kairui Song <kasong@redhat.com>
(cherry picked from commit a60af534132828fd46fa017291f7ed7cfeab1dc3)

Resolves: #1959336

soig committed:
kill bogus comment

should have been killed in commit 4a855e6bebed956c108865823bcc131ebea2841a

(cherry picked from commit a6bc200c28ac841edd12211c96dad8516095a68d)

Resolves: #1959336

ryncsn committed:
99squash: fail early if can't install require modules in initramfs

insmods will consider all arguments as optional modules by default, but
for squash module, all listed modules are required, if any of them is
missing the initramfs won't boot. So pass the '-c' argument to let instmos
know all those modules are mandatory, it will fail and give an error if
it failed to install any of those modules.

(cherry picked from commit 90ac0a2d772ec0d4f90ea16fc5a31be10bdf205e)

Resolves: #1959336

ryncsn committed:
lsinitrd: list squash content as well

When the initramfs is built with squash module enabled, the content
inside squash image will be invisible from the lsinitrd tool. This
commit make lsinitrd detect and list the content of the squash image as
well to avoid possible confusion.

Signed-off-by: Kairui Song <kasong@redhat.com>
(cherry picked from commit 1ff306a3964840fc548188a8aa4862143b347e8a)

Resolves: #1959336

puleglot committed:
busybox: simplify listing of supported utilities

'--list' option is supported since busybox-1.20.0, which was released
in 2010.

(cherry picked from commit 50cc23ba32b0fda63eff7623b529dbeb4e6a38c6)

Resolves: #1959336

ryncsn committed:
99squash: Don't hardcode the squash sub directories

Signed-off-by: Kairui Song <kasong@redhat.com>
(cherry picked from commit 3a2beb037c822e1567f86b63ac24194bffeea991)

Resolves: #1959336

ryncsn committed:
99squash: improve pre-requirements check

Check for systemd-initrd and squashfs-tools in check() to fail early if
it won't work.

Signed-off-by: Kairui Song <kasong@redhat.com>
(cherry picked from commit cfd872392c59c96665c558899880d6b2980065bb)

Resolves: #1959336

ryncsn committed:
99squash: simplify the code

The new dracutsysrootdir could be used to replace the shell function
required_in_root, so drop it and also simplify the code.

Signed-off-by: Kairui Song <kasong@redhat.com>
(cherry picked from commit 4159819fbb20fca8c0a80ddb17e211f481ec7717)

Resolves: #1959336

ryncsn committed:
99squash: Check require module earlier, and properly

Let 99squash fail earlier if required modules are not enabled or
missing, using the new added helper.

Signed-off-by: Kairui Song <kasong@redhat.com>
(cherry picked from commit 83c65fd3db9f5f3436cfe74ec67d2894dedda397)

Resolves: #1959336

ryncsn committed:
fix(squash): this module shouldn't depend on bash

Any other shell should just work.

(cherry picked from commit c552c99ef8e5251e35242e0a3c09eafcd917b31a)

Resolves: #1959336

Commit:
fix(99squash): use kernel config instead of modprobe to check modules

Signed-off-by: Zoltán Böszörményi <zboszor@pr.hu>
(cherry picked from commit 911c2de9c34a2e96be3973954d368d3a14eb6785)

Resolves: #1959336

ryncsn committed:
perf: disable initrd compression when squash module is enabled

With squash module, the initramfs is double compressed, which slow down
the build progress and doesn't shrink the size much.

(cherry picked from commit 7c0bc0b2fd167da42035020dae49af94844f053c)

Resolves: #1959336

Commit:
fix: use find_binary

Use find_binary instead of type -P because the former works if
dracutsysrootdir is set.

Signed-off-by: Zoltán Böszörményi <zboszor@pr.hu>
(cherry picked from commit 433a9ec0147e578b91b4c6e20629a0dbeb9d0725)

Resolves: #1959336

ryncsn committed:
fix(squash): post install should be the last step before stripping

Ensure dracut squash module doesn't effect other steps, and stripping
can cover the new binaries installed by it.

(cherry picked from commit 8c8aecdc63c9389038e78ee712d4809e49add5e1)

Cherry-picked from: 8c8aecdc
Resolves: #1959336

haraldh committed:
fix: shellcheck for modules.d/99squash/setup-squash.sh

(cherry picked from commit 836e2ffac6695201e334c8ef13488e2492b58db0)

Resolves: #1959336

haraldh committed:
fix: shellcheck for modules.d/99squash/clear-squash.sh

(cherry picked from commit 73f5cc7c3754f9561abe948b4b080a84b34c7d1a)

Resolves: #1959336

haraldh committed:
fix: shellcheck for modules.d/99squash/module-setup.sh

(cherry picked from commit 6510ab3f9d6ce555435cf57b716702b5b7663772)

Resolves: #1959336

ryncsn committed:
feat(squash): use busybox for early setup if available

Use busybox can help reduce the size of early setup environment.

With this change, everything is packed in the squash image, and
the setup files will be dropped once squash image setup is done,
so initramfs stage memory usage is reduced to the minimun,
and initramfs decompress is also faster.

File layout of a squash initramfs looks like this:

========================================================================
drwxr-xr-x   1 root     root            0 Feb 15 14:07 .
-rwxr-xr-x   1 root     root          946 Feb 15 14:07 init
lrwxrwxrwx   1 root     root            7 Feb 15 14:07 lib -> usr/lib
drwxr-xr-x   1 root     root            0 Feb 15 14:07 squash
-rw-r--r--   1 root     root     91000832 Feb 15 14:07 squash-root.img
drwxr-xr-x   1 root     root            0 Feb 15 14:07 usr
drwxr-xr-x   1 root     root            0 Feb 15 14:07 usr/bin
-rwxr-xr-x   1 root     root      1293688 Jul 27  2020 usr/bin/busybox
lrwxrwxrwx   1 root     root            7 Feb 15 14:07 usr/bin/echo -> busybox
lrwxrwxrwx   1 root     root            7 Feb 15 14:07 usr/bin/mkdir -> busybox
lrwxrwxrwx   1 root     root            7 Feb 15 14:07 usr/bin/modprobe -> busybox
lrwxrwxrwx   1 root     root            7 Feb 15 14:07 usr/bin/mount -> busybox
lrwxrwxrwx   1 root     root            7 Feb 15 14:07 usr/bin/sh -> busybox
lrwxrwxrwx   1 root     root            7 Feb 15 14:07 usr/bin/switch_root -> busybox
drwxr-xr-x   1 root     root            0 Feb 15 14:07 usr/lib
drwxr-xr-x   1 root     root            0 Feb 15 14:07 usr/lib/dracut
-rw-r--r--   1 root     root           23 Feb 15 14:07 usr/lib/dracut/build-parameter.txt
-rw-r--r--   1 root     root           31 Feb 15 14:07 usr/lib/dracut/dracut-051-93.git20210215.fc33
-rw-r--r--   1 root     root          358 Feb 15 14:07 usr/lib/dracut/modules.txt
-rw-r--r--   1 root     root            0 Feb 15 14:07 usr/lib/dracut/need-initqueue
drwxr-xr-x   1 root     root            0 Feb 15 14:07 usr/lib/modules
drwxr-xr-x   1 root     root            0 Feb 15 14:07 usr/lib/modules/5.10.11-200.fc33.x86_64
drwxr-xr-x   1 root     root            0 Feb 15 14:07 usr/lib/modules/5.10.11-200.fc33.x86_64/kernel
<... kernel module misc files skipped ... >
========================================================================

(cherry picked from commit 90f269f6afe409925bad86f0bd7e9322ad9b4fb0)

Resolves: #1959336

ryncsn committed:
refactor(squash): move the post install scripts into the module-setup.sh

No function change, just move the post install code to 99squash to clean
up dracut.sh.

(cherry picked from commit 95ea16aa606912b7cc8f9942481c8a889b640c15)

Cherry-picked from: 95ea16aa
Resolves: #1959336

ryncsn committed:
refactor(squash): move all setup code to init-squash.sh

Seperate of init.sh and setup-squash.sh is pointless, merge them into one.

(cherry picked from commit f9f4264d038816990aa0a830b31e31d2e8b84cad)

Resolves: #1959336

ryncsn committed:
feat(squash): install and depmod modules seperately

Separately install the modules required for squash image setup.
These modules can be deleted after squash image setup to save
memory.

Signed-off-by: Kairui Song <kasong@redhat.com>
(cherry picked from commit 5a18b24a8b9c20c98f711963ce5407ceb2f3d57b)

Cherry-picked from: 5a18b24a8b9c20c98f711963ce5407ceb2f3d57b
Resolves: #1959336

ryncsn committed:
refactor(squash): don't record mount points in text file

The squasn mount points are recorded in text file so later clean up
script can umount them, this is not needed, the mount points are fixed
so just detect the umount by hardcoded pattern.

(cherry picked from commit bdd194bb8c0427921b87b4dce16c05e18e0c2fb0)

Resolves: #1959336

ryncsn committed:
feat: squash module follow --compress option

Try to make mksquashfs follow --compress option if squash module is
included, if not specified or invalid, fall back to use mksquashfs's
default compressor.

(cherry picked from commit 5d05ffbd87bc27e27f517ebc3454d50729c687e6)

Cherry-picked from: 5d05ffbd
Resolves: #1959336

haraldh committed:
fix(squash): shellcheck for modules.d/99squash

(cherry picked from commit cbef7cf3dac79f014200017e9b13e012f2a53f47)

Resolves: #1959336

ryncsn committed:
fix(squash): don't mount the mount points if already mounted

It is possible that user setup some early mount in the rootfs even
earlier, so just be more robust and cover that case too.

Signed-off-by: Kairui Song <kasong@redhat.com>
(cherry picked from commit 636d6df3134dde1dac72241937724bc59deb9303)

Resolves: #1959336

ryncsn committed:
fix(squash): create relative symlinks

Don't use absolute path, or it may resolve to wrong files after unpack
the initramfs to a sub path on a running system.

Signed-off-by: Kairui Song <kasong@redhat.com>
(cherry picked from commit a2b6be44792b68218e3378a7d844b0f8527a4805)

Resolves: #1959336

hbathini committed:
fix(dracut.sh): handle '-i' option to include files beginning with '.'

While including a directory using '--include' option, the file and
subdirectory names that begin with '.' are not included. Also, dracut
throws a warning message when a subdirectory is empty or only has
files or subdirectories that begin with '.'.

For example, while trying to include /tmpdata directory with the
below tree:

  # tree -a /tmpdata
  /tmpdata
  ├── .anothertestdir
  ├── testdir
  │   └── .testsubdir
  └── .testfile

dracut throws the below warning message:

  # dracut --include /tmpdata /root
  cp: cannot stat '/tmpdata/testdir/*': No such file or directory
  #

and this is how the included /tmpdata directory tree looks:

  # tree -a root
  root
  └── testdir

No file or directory beginning with '.' is included & also, copying
/tmpdata/testdir reported "No such file or directory" warning. Using
'.' instead of '*' in the below command will fix the warning whether
the directory being copied is empty or only has files or directories
that begin with dot:

  $DRACUT_CP -t "$object_destdir" "$dracutsysrootdir$objectname"/*

Also, enable 'dotglob' temporarily to include files and directories
beginning with a `.' in the results of pathname expansion of source
directory being included.

Signed-off-by: Hari Bathini <hbathini@linux.ibm.com>
(cherry picked from commit f1138012c9dc44e6614466c0a8e929fc55e4a5dd)

Cherry-picked from: f1138012
Resolves: #1959336

ryncsn committed:
fix(dracut.sh): handle symlinks appropriately while using '-i' option

[[ -d $symlink ]] will return true if the symlink points to a directory.
So the symlink will not be copied, instead a directory is created with
the symlink name and the content is copied.

Signed-off-by: Kairui Song <kasong@redhat.com>
(cherry picked from commit c7fbc0c8901917baf0d1f0822568e65c6ec00d18)

Cherry-picked from: c7fbc0c8
Resolves: #1959336

ryncsn committed:
feat(squash): install umount util

Also install umount binary, make it possible to cleanup squash overlay
mounts. This is useful for other tools reusing the dracut initramfs built
with squash module enabled.

Signed-off-by: Kairui Song <kasong@redhat.com>
(cherry picked from commit 563f543424c66bf38e6cbd3f489655d45ad9b5c5)

Resolves: #1959336

Commit:
fix(squash): keep ld cache under initdir

When running kdump on PowerPC, the following bug is hit:
[    0.391629] Freeing unused kernel memory: 5568K
[    0.391634] This architecture does not have kernel memory protection.
/bin/sh: error while loading shared libraries: libc.so.6: cannot open shared object file: No such file or directory
[    0.392214] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00007f00
[    0.392214]
[    0.392223] CPU: 6 PID: 1 Comm: init Not tainted 4.18.0-319.el8.ppc64le #1
[    0.392228] Call Trace:
[    0.392234] [c00000000c703c10] [c000000008ecb94c] dump_stack+0xb0/0xf4 (unreliable)
[    0.392243] [c00000000c703c50] [c000000008167324] panic+0x148/0x3c4
[    0.392249] [c00000000c703cf0] [c000000008170474] do_exit+0xcd4/0xd40
[    0.392255] [c00000000c703dc0] [c0000000081705b0] do_group_exit+0x60/0x110
[    0.392261] [c00000000c703e00] [c000000008170684] sys_exit_group+0x24/0x30
[    0.392268] [c00000000c703e20] [c00000000800b408] system_call+0x5c/0x70

This is due to the non-conventional library path:
ldd /bin/bash
        linux-vdso64.so.1 (0x00007fffbdc90000)
        libtinfo.so.6 => /lib64/libtinfo.so.6 (0x00007fffbda80000)
        libdl.so.2 => /lib64/libdl.so.2 (0x00007fffbda50000)
        libc.so.6 => /lib64/glibc-hwcaps/power9/libc-2.28.so (0x00007fffbd830000)
                            ^^^
        /lib64/ld64.so.2 (0x00007fffbdcb0000)

ldd finds the path by libc.so.6 -> /usr/lib64/libc.so.6 ->
/usr/lib64/glibc-hwcaps/power9/libc-2.28.so and cache the result. So
when dracut_install, it only saw
'/usr/lib64/glibc-hwcaps/power9/libc-2.28.so' and blind to
'/usr/lib64/libc.so.6'

In the final kdumpimg, the symlink /usr/lib64/libc.so.6 is not created,
hence ld.so can not find the /usr/lib64/glibc-hwcaps/power9/libc-2.28.so

On the other hand, during the process of building kdumpimg, all of dynamic library info
have been cached in ld.so.cache. Hence this bug can be simplely resolved
by keeping ld cache under $initdir/etc.

Signed-off-by: Pingfan Liu <piliu@redhat.com>
---
v3 -> v4:
  use inst() instead of copy
v2 -> v3:
  fix format by shfmt -s -w modules.d/99squash/module-setup.sh
v1 -> v2:
  cp -r /etc/ld.so* instead of move, since after switch-root, initdir
can not be seen any longer

(cherry picked from commit dc21638c3f0acbb54417f3bfb6294ad5514bf2db)

Resolves: #1959336

lnykryn committed:
fix: revise squash module checks

Based on 16efdfa25c65c31acc97d91b7baf4f7bb4ea8510

lnykryn committed:
refactor(squash): structure in a cleaner way

Simplify the squash mount layout. Instead of overlay on each top
directory (/etc, /usr), just mount and switch_root into the squash
image, with a overlay on top of it.

Also install the binaries and setup scripts separately, so the squash
setup code and the squash image content is independent of each other,
all squash setup script and binaries can be deleted safely upon
switch_root.

With this change, previous squash clean up service and other tricky
implementations are all gone.

This commit depends on systemd commits from:
https://github.com/systemd/systemd/pull/18124

Previouly systemd doesn't recognize non-ramfs initramfs, now this is
doable with SYSTEMD_IN_INITRD=lenient

Signed-off-by: Kairui Song <kasong@redhat.com>
(cherry picked from commit 8f7c332e488f88e5845a3c7954af7934c2f1e37b)

Cherry-picked from: 8f7c332e
Resolves: #1959336

ryncsn committed:
fix(squash): remove tailing '/' when installing ld.so.conf.d

This tailing '/' will result in following error:

dracut-install: ERROR: installing '/etc/ld.so.conf.d/'
dracut: FAILED: /usr/lib/dracut/dracut-install -D /var/tmp/dracut.kEFQLs/initramfs -d /etc/ld.so.conf.d/

Signed-off-by: Kairui Song <kasong@redhat.com>

Related:#1959336

ryncsn commented:
The issue with dracutsysrootdir can be resolved with:
```
diff --git a/modules.d/99squash/module-setup.sh b/modules.d/99squash/module-setup.sh
index 14790ce8..11964b97 100644
--- a/modules.d/99squash/module-setup.sh
+++ b/modules.d/99squash/module-setup.sh
@@ -39,9 +39,10 @@ installpost() {
 
     # Copy dracut spec files out side of the squash image
     # so dracut rebuild and lsinitrd can work
+    mkdir -p "$initdir/usr/lib/dracut/"
     for file in "$squash_dir"/usr/lib/dracut/*; do
         [[ -f $file ]] || continue
-        DRACUT_RESOLVE_DEPS=1 dracutsysrootdir="$squash_dir" inst "${file#$squash_dir}"
+        cp "$file" "$initdir/${file#$squash_dir}"
     done
```

And need to move the $compress args parsing to later part after mksquashfs, so mksquashfs can use $compress as the original parameter value:

```
diff --git a/dracut.sh b/dracut.sh
index fa14e3ce..e559bb96 100755
--- a/dracut.sh
+++ b/dracut.sh
@@ -825,53 +825,6 @@ if [[ $_no_compress_l = "cat" ]]; then
     compress="cat"
 fi
 
-if ! [[ $compress ]]; then
-    # check all known compressors, if none specified
-    for i in pigz gzip lz4 lzop zstd lzma xz lbzip2 bzip2 cat; do
-        command -v "$i" &>/dev/null || continue
-        compress="$i"
-        break
-    done
-    if [[ $compress = cat ]]; then
-            printf "%s\n" "dracut: no compression tool available. Initramfs image is going to be big." >&2
-    fi
-fi
-
-# choose the right arguments for the compressor
-case $compress in
-    bzip2|lbzip2)
-        if [[ "$compress" =  lbzip2 ]] || command -v lbzip2 &>/dev/null; then
-            compress="lbzip2 -9"
-        else
-            compress="bzip2 -9"
-        fi
-        ;;
-    lzma)
-        compress="lzma -9 -T0"
-        ;;
-    xz)
-        compress="xz --check=crc32 --lzma2=dict=1MiB -T0"
-        ;;
-    gzip|pigz)
-        if [[ "$compress" = pigz ]] || command -v pigz &>/dev/null; then
-            compress="pigz -9 -n -T -R"
-        elif command -v gzip &>/dev/null && gzip --help 2>&1 | grep -q rsyncable; then
-            compress="gzip -n -9 --rsyncable"
-        else
-            compress="gzip -n -9"
-        fi
-        ;;
-    lzo|lzop)
-        compress="lzop -9"
-        ;;
-    lz4)
-        compress="lz4 -l -9"
-        ;;
-    zstd)
-       compress="zstd -15 -q -T0"
-       ;;
-esac
-
 [[ $hostonly = yes ]] && hostonly="-h"
 [[ $hostonly != "-h" ]] && unset hostonly
 
@@ -1848,6 +1801,53 @@ if [[ $create_early_cpio = yes ]]; then
     fi
 fi
 
+if ! [[ $compress ]]; then
+    # check all known compressors, if none specified
+    for i in pigz gzip lz4 lzop zstd lzma xz lbzip2 bzip2 cat; do
+        command -v "$i" &>/dev/null || continue
+        compress="$i"
+        break
+    done
+    if [[ $compress = cat ]]; then
+            printf "%s\n" "dracut: no compression tool available. Initramfs image is going to be big." >&2
+    fi
+fi
+
+# choose the right arguments for the compressor
+case $compress in
+    bzip2|lbzip2)
+        if [[ "$compress" =  lbzip2 ]] || command -v lbzip2 &>/dev/null; then
+            compress="lbzip2 -9"
+        else
+            compress="bzip2 -9"
+        fi
+        ;;
+    lzma)
+        compress="lzma -9 -T0"
+        ;;
+    xz)
+        compress="xz --check=crc32 --lzma2=dict=1MiB -T0"
+        ;;
+    gzip|pigz)
+        if [[ "$compress" = pigz ]] || command -v pigz &>/dev/null; then
+            compress="pigz -9 -n -T -R"
+        elif command -v gzip &>/dev/null && gzip --help 2>&1 | grep -q rsyncable; then
+            compress="gzip -n -9 --rsyncable"
+        else
+            compress="gzip -n -9"
+        fi
+        ;;
+    lzo|lzop)
+        compress="lzop -9"
+        ;;
+    lz4)
+        compress="lz4 -l -9"
+        ;;
+    zstd)
+       compress="zstd -15 -q -T0"
+       ;;
+esac
+
 if ! (
         umask 077; cd "$initdir"
         find . -print0 | sort -z \
```

Everything else looks OK.

ryncsn commented:
Everything else seems good, just found an error message:

    dracut-install: ERROR: installing '/etc/ld.so.conf.d/'
    dracut: FAILED: /usr/lib/dracut/dracut-install -D /var/tmp/dracut.kEFQLs/initramfs -d /etc/ld.so.conf.d/


In most case it should be totally fine, ld.so.conf.d isn't very useful for early squash loader, it was installed just in case, and so far everything still works if it's absent.

I've made a PR to fix that, will be great if this can be also backported:
https://github.com/dracutdevs/dracut/pull/1574/files

lnykryn committed:
dracut-functions: backport block_is_* functions

Based on 480aa9695f8c2e2b30c8f41ae8483140020d23db
RHEL-only

Related: #1959336

ryncsn committed:
fix(squash): fixes related to squash module rebase

fix fro missing dracutsysrootdir
move the $compress args parsing to later part after mksquashfs, so mksquashfs can use $compress as the original parameter value

Related:#1959336

lnykryn commented:
Thanks for the review, I've included all your suggestions.
